src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/Int32/ClosedRangeInt32FrozenDictionary.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Frozen
+{
+    /// <summary>Provides a frozen dictionary to use when the key is an <see cref="int"/>, the default comparer is used, and the item count is small.</summary>
+    /// <remarks>
+    /// No hashing here, just a simple range check instead.
+    /// </remarks>
+    internal sealed class ClosedRangeInt32FrozenDictionary<TValue> : FrozenDictionary<int, TValue>
+    {
+        private readonly int[] _keys;
+        private readonly TValue[] _values;
+        private readonly int _min;
+
+        // assumes the keys are sorted
+        internal ClosedRangeInt32FrozenDictionary(int[] keys, TValue[] values)
+            : base(EqualityComparer<int>.Default)
+        {
+            Debug.Assert(keys.Length != 0);
+
+            _keys = keys;
+            _values = values;
+            _min = keys[0];
+        }
+
+        private protected override int[] KeysCore => _keys;
+        private protected override TValue[] ValuesCore => _values;
+        private protected override Enumerator GetEnumeratorCore() => new Enumerator(_keys, _values);
+        private protected override int CountCore => _keys.Length;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private protected override ref readonly TValue GetValueRefOrNullRefCore(int key)
+        {
+            TValue[] values = _values;
+            if ((uint)(key - _min) < (uint)values.Length)
+            {
+                return ref values[key - _min];
+            }
```

Comment:
For the bounds check to actually be eliminated, I expect you need to hoist the subtraction as well; with the field access, I don't believe common subexpression elimination will kick in.
```suggestion
            TValue[] values = _values;
            int i = key - _min;
            if ((uint)i < (uint)values.Length)
            {
                return ref values[diff];
            }
```

Reply:
Done.